infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/StateTransfer.java

```diff
@@ -30,96 +33,197 @@ private StateTransfer() {
+                    // Fetch all missing entries in this range [chunkStart - chunkEnd].
+                    List<Long> chunk = getMissingEntriesChunk(layout, runtime, endpoint,
+                            chunkStart, chunkEnd);
+                    if (chunk.isEmpty()) {
+                        continue;
+                    }
 
-            List<LogData> entries = new ArrayList<>();
-            for (long x = chunkStart; x <= chunkEnd; x++) {
-                if (!dataMap.containsKey(x)) {
-                    log.error("Missing address {} in range {}-{}", x, chunkStart, chunkEnd);
-                    throw new IllegalStateException("Missing address");
+                    // Read and write in chunks of chunkSize.
+                    transferChunk(layout, runtime, endpoint, chunk);
                 }
-                entries.add((LogData) dataMap.get(x));
-            }
+            } catch (OverwriteException oe) {
 
-            for (String endpoint : endpoints) {
-                // Write segment chunk to the new logunit
-                ts1 = System.currentTimeMillis();
-                boolean transferSuccess = runtime.getLayoutView().getRuntimeLayout(layout)
-                        .getLogUnitClient(endpoint)
-                        .writeRange(entries).get();
-                ts2 = System.currentTimeMillis();
-
-                if (!transferSuccess) {
-                    log.error("stateTransfer: Failed to transfer {}-{} to {}", chunkStart,
-                            chunkEnd, endpoint);
-                    throw new IllegalStateException("Failed to transfer!");
+                if (overwriteRetries.getAndIncrement() >= OVERWRITE_RETRIES) {
+                    throw new RetryExhaustedException("StateTransfer: Retries exhausted.");
                 }
+                throw new RetryNeededException();
+            }
+
+            return true;
+        }).setOptions(retry -> {
+            retry.setMaxRetryThreshold(MAX_RETRY_TIMEOUT);
+            retry.setRandomPortion(RANDOM_FACTOR_BACKOFF);
+        }).run();
+    }
+
+    /**
+     * Send the trimMark to the new/healing nodes.
+     * If this times out or fails, the Action performing the stateTransfer
+     * fails and retries.
+     * TrimMark is the first address present on the log unit server.
+     * Perform the prefix trim on the preceding address = (trimMark - 1).
+     * Since the LU will reject trim decisions made from older epochs, we
+     * need to adjust the new trim mark to have the new layout's epoch.
+     *
+     * @param layout   Current layout.
+     * @param runtime  Corfu runtime instance.
+     * @param endpoint Endpoint ot transfer data to.
+     * @return Trim Address.
+     */
+    private static long setTrimOnNewLogUnit(Layout layout, CorfuRuntime runtime,
+                                            String endpoint) {
+
+        long trimMark = runtime.getAddressSpaceView().getTrimMark().getSequence();
```

Comment:
This can execute on a different layout, but subsequent calls don't. Why not call this with a runtime layout?

Also, what happens if there are other intermittent failures other than WrongEpochException ?

Reply:
The reads are on the AddressSpaceView layer to abstract the replication layer and also the exceptions are taken care of.
This exposes a TrimException on which the action fails and is retried with the new trim mark.
WrongEpochExceptions should be just retried. Similarly, NetworkExceptions, Timeouts.
The writes do not use this layoutHelper method and hence are aborted if any exception is hit.
I haven't modified this functionality in this PR. I will address that in the next PR with appropriate testing.